README.md
# Conditions
![Build Status](https://github.com/johnallen3d/conditions/actions/workflows/ci.yml/badge.svg) [![Coverage Status](https://coveralls.io/repos/github/johnallen3d/conditions/badge.svg?branch=feature/track-coverage)](https://coveralls.io/github/johnallen3d/conditions?branch=feature/track-coverage) ![Crate Status](https://img.shields.io/crates/v/conditions.svg)

A command line tool for getting the weather conditions at the current location.

## Weather Sources

- [open-meteo.com](https://open-meteo.com/)
- [weatherapi.com](https://www.weatherapi.com/) \*

\* requires an [api key](https://www.weatherapi.com/docs/#intro-authentication)

## Installation

At this time it is necessary to compile and install the crate locally. The simplest way to do this is to [install the Rust toolchain](https://rustup.rs/).

```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

Then use `cargo` to build and install from [crates.io](https://crates.io/crates/conditions)

```bash
cargo install conditions
```

## Usage

By default the Open-Meteo weather provider will be used along with a location determined by the machine's ip address.

```sh
> conditions current | jq
{
  "temp": 57,
  "icon": ""
}
```

### Weather API

In order to use the Weather API provider create an account and then an API Key. Then persist the api key in settings.

```bash
conditions weather-api-key set [your-api-key]
```

### Location

If you prefer to set a specific location you can do so via a postal code:

```bash
conditions location set "[postal-code], [country]"
```

For example:

```bash
conditions location set "10001, usa"
```

### SketchyBar

[SketchyBar](https://github.com/FelixKratz/SketchyBar) is a very customizable bar app for macOS that is deliberately light on resource usage (written in C vs more common WebView approach). Widgets in SketchyBar are refreshed (if need be) via "plugins" (typically shell scripts).

#### Manual

Here's how I'm using `conditions` with SketchyBar in the more traditional plugin approach.

```bash
#!/bin/bash

conditions="$(conditions current)"
icon=$(echo "$conditions" | jq -r .icon)
temp=$(echo "$conditions" | jq -r .temp)

sketchybar -m \
  --set weather_logo icon="${icon}" \
  --set weather label="${temp}°F"
```

#### Direct Integration

Alternatively `conditions` can directly update the running instance of SketchyBar without making calls from a shell script using `sketchybar -m`. In order to use this experimental version, the `sketchybar` feature must be enabled on install.

```bash
cargo install conditions --features sketchybar
```

Then when setting up a weather widget in `sketchybarrc`, `conditions` can be invoked directly.

```bash
sketchybar \
  --add item weather right \
  update_freq=1800 \
  icon.drawing=off \
  script="$HOME/.cargo/bin/conditions current --icon="

sketchybar \
  --add item weather_logo right \
  icon.font="Hack Nerd Font:Regular:14.0" \
  label.drawing=off \
```

Note, at this time `conditions` is hard-coded with widgets named `weather` and `weather_logo`:

```rust
let message = format!(
    "--set weather_logo icon=\"{}\" --set weather label=\"{}°{}\"",
    conditions.icon,
    conditions.temp_f as i32,
    self.config.unit.as_char().to_ascii_uppercase()
);
```

In my bar this results in:

![example SketchyBar](./assets/sketchybar-example.png)

## Tasks

Run tasks from this directory via: `xc [task-name]`

### check

```sh
cargo build
```

### build

```sh
cargo build
```

### run

```sh
cargo run
```

### install

```sh
cargo install --path .
```

## Dependencies

| Project             | Version |
| ------------------- | ------- |
| rust-lang.org       | ^1.6    |
| rust-lang.org/cargo | ^0.66   |
| git-scm.org         | ^2.38   |
| xcfile.dev          | ^0      |

## Why

I wanted to learn rust and had a [(cargo-culted) shell script](https://github.com/johnallen3d/dotfiles/blob/16054c903bc8cc0ca939c279382ec6b15eb1bc7c/dot_config/sketchybar/plugins/executable_weather.sh#L1) that performed this task for usage with [sketchybar](https://github.com/FelixKratz/SketchyBar).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0](https://github.com/johnallen3d/conditions/compare/v0.1.3...v0.2.0) - 2023-10-08

### Added
- *(current)* allow optional region ([#125](https://github.com/johnallen3d/conditions/pull/125))
- *(location)* add caching ([#123](https://github.com/johnallen3d/conditions/pull/123))

### Other
- remove usage of `eprintln!` ([#126](https://github.com/johnallen3d/conditions/pull/126))
- *(deps)* bump ureq from 2.7.1 to 2.8.0
- *(deps)* bump clap from 4.4.5 to 4.4.6 ([#118](https://github.com/johnallen3d/conditions/pull/118))
- *(deps)* bump thiserror from 1.0.48 to 1.0.49 ([#117](https://github.com/johnallen3d/conditions/pull/117))
- *(deps)* bump clap from 4.4.4 to 4.4.5 ([#116](https://github.com/johnallen3d/conditions/pull/116))
- make clippy enforce pedantic lints ([#115](https://github.com/johnallen3d/conditions/pull/115))
- *(deps)* bump clap from 4.4.3 to 4.4.4 ([#114](https://github.com/johnallen3d/conditions/pull/114))
- *(deps)* bump clap from 4.4.2 to 4.4.3 ([#111](https://github.com/johnallen3d/conditions/pull/111))
- always squash the merges!
- skip coverage for dependabot
- add a crates.io badge ([#110](https://github.com/johnallen3d/conditions/pull/110))
- improve unit test coverage ([#107](https://github.com/johnallen3d/conditions/pull/107))
- add code coverage tracking ([#108](https://github.com/johnallen3d/conditions/pull/108))
- prevent release-plz running when dependabot ([#104](https://github.com/johnallen3d/conditions/pull/104))
- *(deps)* bump tibdex/github-app-token from 1 to 2 ([#102](https://github.com/johnallen3d/conditions/pull/102))
- *(deps)* bump actions/checkout from 3 to 4 ([#101](https://github.com/johnallen3d/conditions/pull/101))
- *(deps)* bump rustls-webpki from 0.100.1 to 0.100.2 ([#97](https://github.com/johnallen3d/conditions/pull/97))
- configure dependabot for actions ([#100](https://github.com/johnallen3d/conditions/pull/100))
- run dependabot checks daily
- fix dependabot auto merge
- enable dependabot auto merge ([#96](https://github.com/johnallen3d/conditions/pull/96))
- *(deps)* bump serde from 1.0.179 to 1.0.188 ([#94](https://github.com/johnallen3d/conditions/pull/94))
- towards (more) simplicity ([#93](https://github.com/johnallen3d/conditions/pull/93))
- update CI configuration ([#92](https://github.com/johnallen3d/conditions/pull/92))
- extract an http client trait ([#85](https://github.com/johnallen3d/conditions/pull/85))

## [0.1.3](https://github.com/johnallen3d/conditions/compare/v0.1.2...v0.1.3) - 2023-09-06

### Fixed
- *(weather)* typo and invalid enum ([#87](https://github.com/johnallen3d/conditions/pull/87))

## [0.1.2](https://github.com/johnallen3d/conditions/compare/v0.1.1...v0.1.2) - 2023-09-05

### Added
- *(weather)* add fallback ([#83](https://github.com/johnallen3d/conditions/pull/83))

### Other
- add installation instructions ([#81](https://github.com/johnallen3d/conditions/pull/81))

## [0.1.1](https://github.com/johnallen3d/conditions/compare/v0.1.0...v0.1.1) - 2023-09-04

### Other
- *(bin)* setup binary release ([#77](https://github.com/johnallen3d/conditions/pull/77))
- *(deps)* bump thiserror from 1.0.47 to 1.0.48 ([#70](https://github.com/johnallen3d/conditions/pull/70))

## [0.1.0](https://github.com/johnallen3d/conditions/releases/tag/v0.1.0) - 2023-02-13

### Added
- *(config)* add commands to view config info
- allow user to set temp unit (#19)
- allow user to store location (#17)
- allow user to view stored token (#16)
- [**breaking**] add support for setting token (#12)
- read weatherapi token from config (#9)
- *(bin)* initial commit

### Fixed
- disambiguate the term `conditions`

### Other
- *(deps)* bump clap from 3.2.23 to 4.1.4
- add dependabot configuration
- move from release-please to release-plz
- better help for unit set
- _slightly_ better error handling (#23)
- *(unit)* use enum for unit values
- move current function to struct
- *(config)* move to module
- fix release-please permissions (#10)
- add GitHub Actions configuration (#6)
- add a license
- link to shell script inspiration
- *(usage)* add a note about usage with SketchyBar
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


src/icons.rs
use lazy_static::lazy_static;
use std::collections::HashMap;

use crate::weather::Provider;

#[derive(Debug, PartialEq)]
pub enum TimeOfDay {
    Night,
    Day,
}

impl From<u8> for TimeOfDay {
    fn from(is_day: u8) -> Self {
        match is_day {
            1 => TimeOfDay::Day,
            _ => TimeOfDay::Night,
        }
    }
}

impl TimeOfDay {
    pub fn icon(
        &self,
        provider: crate::weather::Provider,
        code: i32,
    ) -> String {
        let icons: &HashMap<i32, &'static str> = match provider {
            Provider::WeatherAPI => match self {
                TimeOfDay::Day => &WEATHERAPI_DAY_ICONS,
                TimeOfDay::Night => &WEATHERAPI_NIGHT_ICONS,
            },
            Provider::OpenMeteo => match self {
                TimeOfDay::Day => &OPEN_METEO_DAY_ICONS,
                TimeOfDay::Night => &OPEN_METEO_NIGHT_ICONS,
            },
        };

        icons.get(&code).unwrap_or(&"?").to_string()
    }
}

lazy_static! {
    static ref WEATHERAPI_DAY_ICONS: HashMap<i32, &'static str> = {
        let mut m = HashMap::new();
        m.insert(1000, " "); // Clear/113
        m.insert(1003, " "); // Partly cloudy/116
        m.insert(1006, " "); // Cloudy/119
        m.insert(1009, " "); // Overcast/122
        m.insert(1030, " "); // Mist/143
        m.insert(1063, " "); // Patchy rain possible/176
        m.insert(1066, " "); // Patchy snow possible/179
        m.insert(1069, " "); // Patchy sleet possible/182
        m.insert(1072, " "); // Patchy freezing drizzle possible/185
        m.insert(1087, " "); // Thundery outbreaks possible/200
        m.insert(1114, " "); // Blowing snow/227
        m.insert(1117, " "); // Blizzard/230
        m.insert(1135, " "); // Fog/248
        m.insert(1147, " "); // Freezing fog/260
        m.insert(1150, " "); // Patchy light drizzle/263
        m.insert(1153, " "); // Light drizzle/266
        m.insert(1168, " "); // Freezing drizzle/281
        m.insert(1171, " "); // Heavy freezing drizzle/284
        m.insert(1180, " "); // Patchy light rain/293
        m.insert(1183, " "); // Light rain/296
        m.insert(1186, " "); // Moderate rain at times/299
        m.insert(1189, " "); // Moderate rain/302
        m.insert(1192, " "); // Heavy rain at times/305
        m.insert(1195, " "); // Heavy rain/308
        m.insert(1198, " "); // Light freezing rain/311
        m.insert(1201, " "); // Moderate or heavy freezing rain/314
        m.insert(1204, " "); // Light sleet/317
        m.insert(1207, " "); // Moderate or heavy sleet/320
        m.insert(1210, " "); // Patchy light snow/323
        m.insert(1213, " "); // Light snow/326
        m.insert(1216, " "); // Patchy moderate snow/329
        m.insert(1219, " "); // Moderate snow/332
        m.insert(1222, " "); // Patchy heavy snow/335
        m.insert(1225, " "); // Heavy snow/338
        m.insert(1237, " "); // Ice pellets/350
        m.insert(1240, " "); // Light rain shower/353
        m.insert(1243, " "); // Moderate or heavy rain shower/356
        m.insert(1246, " "); // Torrential rain shower/359
        m.insert(1249, " "); // Light sleet showers/362
        m.insert(1252, " "); // Moderate or heavy sleet showers/365
        m.insert(1255, " "); // Light snow showers/368
        m.insert(1258, " "); // Moderate or heavy snow showers/371
        m.insert(1261, " "); // Light showers of ice pellets/374
        m.insert(1264, " "); // Moderate or heavy showers of ice pellets/377
        m.insert(1273, " "); // Patchy light rain with thunder/386
        m.insert(1276, " "); // Moderate or heavy rain with thunder/389
        m.insert(1279, " "); // Patchy light snow with thunder/392
        m.insert(1282, " "); // Moderate or heavy snow with thunder/395
        m
    };

    static ref WEATHERAPI_NIGHT_ICONS: HashMap<i32, &'static str> = {
        let mut m = HashMap::new();
        m.insert(1000, ""); // Clear/113
        m.insert(1003, ""); // Partly cloudy/116
        m.insert(1006, ""); // Cloudy/119
        m.insert(1009, ""); // Overcast/122
        m.insert(1030, ""); // Mist/143
        m.insert(1063, ""); // Patchy rain possible/176
        m.insert(1066, ""); // Patchy snow possible/179
        m.insert(1069, ""); // Patchy sleet possible/182
        m.insert(1072, ""); // Patchy freezing drizzle possible/185
        m.insert(1087, ""); // Thundery outbreaks possible/200
        m.insert(1114, ""); // Blowing snow/227
        m.insert(1117, ""); // Blizzard/230
        m.insert(1135, ""); // Fog/248
        m.insert(1147, ""); // Freezing fog/260
        m.insert(1150, ""); // Patchy light drizzle/263
        m.insert(1153, ""); // Light drizzle/266
        m.insert(1168, ""); // Freezing drizzle/281
        m.insert(1171, ""); // Heavy freezing drizzle/284
        m.insert(1180, ""); // Patchy light rain/293
        m.insert(1183, ""); // Light rain/296
        m.insert(1186, ""); // Moderate rain at times/299
        m.insert(1189, ""); // Moderate rain/302
        m.insert(1192, ""); // Heavy rain at times/305
        m.insert(1195, ""); // Heavy rain/308
        m.insert(1198, ""); // Light freezing rain/311
        m.insert(1201, ""); // Moderate or heavy freezing rain/314
        m.insert(1204, ""); // Light sleet/317
        m.insert(1207, ""); // Moderate or heavy sleet/320
        m.insert(1210, ""); // Patchy light snow/323
        m.insert(1213, ""); // Light snow/326
        m.insert(1216, ""); // Patchy moderate snow/329
        m.insert(1219, ""); // Moderate snow/332
        m.insert(1222, ""); // Patchy heavy snow/335
        m.insert(1225, ""); // Heavy snow/338
        m.insert(1237, ""); // Ice pellets/350
        m.insert(1240, ""); // Light rain shower/353
        m.insert(1243, ""); // Moderate or heavy rain shower/356
        m.insert(1246, ""); // Torrential rain shower/359
        m.insert(1249, ""); // Light sleet showers/362
        m.insert(1252, ""); // Moderate or heavy sleet showers/365
        m.insert(1255, ""); // Light snow showers/368
        m.insert(1258, ""); // Moderate or heavy snow showers/371
        m.insert(1261, ""); // Light showers of ice pellets/374
        m.insert(1264, ""); // Moderate or heavy showers of ice pellets/377
        m.insert(1273, ""); // Patchy light rain with thunder/386
        m.insert(1276, ""); // Moderate or heavy rain with thunder/389
        m.insert(1279, ""); // Patchy light snow with thunder/392
        m.insert(1282, ""); // Moderate or heavy snow with thunder/395
        m
    };

    static ref OPEN_METEO_DAY_ICONS: HashMap<i32, &'static str> = {
        let mut m = HashMap::new();
        m.insert(0, " "); // Clear sky
        m.insert(1, " ");
        m.insert(2, " ");
        m.insert(3, " "); // Mainly clear, partly cloudy, and overcast
        m.insert(45, " ");
        m.insert(48, " "); // Fog
        m.insert(51, " ");
        m.insert(53, " ");
        m.insert(55, " "); // Drizzle: Light, moderate, and dense intensity
        m.insert(56, " ");
        m.insert(57, " "); // Freezing Drizzle: Light and dense intensity
        m.insert(61, " ");
        m.insert(63, " ");
        m.insert(65, " "); // Rain: Slight, moderate and heavy intensity
        m.insert(66, " ");
        m.insert(67, " "); // Freezing Rain: Light and heavy intensity
        m.insert(71, " ");
        m.insert(73, " ");
        m.insert(75, " "); // Snow fall: Slight, moderate, and heavy intensity
        m.insert(77, " "); // Snow grains
        m.insert(80, " ");
        m.insert(81, " ");
        m.insert(82, " "); // Rain showers: Slight, moderate, and violent
        m.insert(85, " ");
        m.insert(86, " "); // Snow showers slight and heavy
        m.insert(95, " "); // Thunderstorm: Slight or moderate
        m.insert(96, " ");
        m.insert(99, " "); // Thunderstorm with slight and heavy hail
        m
    };

    static ref OPEN_METEO_NIGHT_ICONS: HashMap<i32, &'static str> = {
        let mut m = HashMap::new();
        m.insert(0, ""); // Clear sky
        m.insert(1, "");
        m.insert(2, "");
        m.insert(3, ""); // Mainly clear, partly cloudy, and overcast
        m.insert(45, "");
        m.insert(48, ""); // Fog
        m.insert(51, "");
        m.insert(53, "");
        m.insert(55, ""); // Drizzle: Light, moderate, and dense intensity
        m.insert(56, "");
        m.insert(57, ""); // Freezing Drizzle: Light and dense intensity
        m.insert(61, "");
        m.insert(63, "");
        m.insert(65, ""); // Rain: Slight, moderate and heavy intensity
        m.insert(66, "");
        m.insert(67, ""); // Freezing Rain: Light and heavy intensity
        m.insert(71, "");
        m.insert(73, "");
        m.insert(75, ""); // Snow fall: Slight, moderate, and heavy intensity
        m.insert(77, ""); // Snow grains
        m.insert(80, "");
        m.insert(81, "");
        m.insert(82, ""); // Rain showers: Slight, moderate, and violent
        m.insert(85, "");
        m.insert(86, ""); // Snow showers slight and heavy
        m.insert(95, ""); // Thunderstorm: Slight or moderate
        m.insert(96, "");
        m.insert(99, ""); // Thunderstorm with slight and heavy hail
        m
    };
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_converts_1_to_day() {
        assert_eq!(TimeOfDay::from(1), TimeOfDay::Day);
    }

    #[test]
    fn it_converts_0_to_night() {
        assert_eq!(TimeOfDay::from(0), TimeOfDay::Night);
    }

    #[test]
    fn it_converts_any_other_value_to_night() {
        assert_eq!(TimeOfDay::from(42), TimeOfDay::Night);
    }

    #[test]
    fn valid_code_for_day() {
        let icon =
            TimeOfDay::Day.icon(crate::weather::Provider::WeatherAPI, 1006);

        assert_eq!(icon, " ".to_string());
    }

    #[test]
    fn valid_code_for_night() {
        let icon =
            TimeOfDay::Night.icon(crate::weather::Provider::OpenMeteo, 71);

        assert_eq!(icon, "".to_string());
    }

    #[test]
    fn invalid_code_for() {
        let icon =
            TimeOfDay::Night.icon(crate::weather::Provider::WeatherAPI, 9999);

        assert_eq!(icon, "?".to_string());
    }
}
